lib/src/main/java/com/segment/analytics/kotlin/destinations/consent/ConsentManager.kt

```diff
@@ -0,0 +1,181 @@
+        settings.integrations.forEach { integrationName, integrationJson ->
+            // If the integration has the consent key:
+            integrationJson.jsonObject[CONSENT_SETTINGS_KEY]?.let {
+
+                // Build list of categories required for this integration
+                val categories: MutableList<String> = mutableListOf()
+                (it.jsonObject.get(CATEGORIES_KEY) as JsonArray).forEach { categoryJsonElement ->
+                    categories.add(categoryJsonElement.toString().replace("\"", "").trim())
+                }
+                destinationMapping[integrationName] = categories.toTypedArray()
+            }
+        }
+
+        // Set hasUnmappedDestinations
+        try {
+            settings.toJsonElement().jsonObject.get(CONSENT_SETTINGS_KEY)?.let {
+                val jsonElement = it.jsonObject.get(HAS_UNMAPPED_DESTINATIONS_KEY)
+                println("hasUnmappedDestinations jsonElement: $jsonElement")
+                hasUnmappedDestinations = jsonElement.toString() == "true"
+            }
+        } catch (t: Throwable) {
+            println("Couldn't parse settings object to check for 'hasUnmappedDestinations'")
+        }
+
+        // Set enabledAtSegment
+        try {
+            settings.toJsonElement().jsonObject.get(CONSENT_SETTINGS_KEY)?.let {
+                enabledAtSegment = true
+            }
+        } catch (t: Throwable) {
+            println("Couldn't parse settings object to check if 'enabledAtSegment'.")
+        }
+
+        return ConsentState(destinationMapping, hasUnmappedDestinations, enabledAtSegment)
+    }
+
+    override fun execute(event: BaseEvent): BaseEvent? {
+
+        return if (started.get()) {
+            // Stamp consent on the event
+            stampEvent(event)
+            event
+        } else {
+            queuedEvents?.add(event)
+            null
+        }
+    }
+
+    /**
+     * Add the consent status to the event's context object.
+     */
+    private fun stampEvent(event: BaseEvent) {
+        event.context = buildJsonObject {
+            event.context.forEach { key, json ->
+                put(key, json)
+            }
+            put(CONSENT_KEY, buildJsonObject {
+                put(CATEGORY_PREFERENCE_KEY, buildJsonObject {
+                    val categories = consentProvider.getCategories()
+                    categories.forEach { (category, status) ->
+                        put(category, JsonPrimitive(status))
+                    }
+                })
+            })
+        }
+    }
+
+    /**
+     * Notify the ConsentManagementPlugin that consent has changed. This will
+     * trigger the Segment Consent Preference event to be fired.
+     */
+    fun notifyConsentChanged() {
+        consentChange?.let {
+            it()
+        }
+    }
+
+    fun start() {
+        started.set(true)
+
+        for (event in queuedEvents!!) {
+            analytics?.process(event)
+        }
+
+        queuedEvents?.clear()
```

Comment:
not sure how we gonna use the `start` method, but it's better to make `queuedEvents` a queue instead of list, so we can do
```
while (queueEvents is not empty) {
  poll from queue
  analytics.process
}
```
avoiding to clear at the end, thus avoiding concurrent modification issues.